### ISSUE TITLE
Extend API to allow running collective ops across a subset of ranks (cross, local, etc.)

### ISSUE THREAD

**Issue by tgaddair** [horovod/horovod#2139]
Currently, Horovod's API is restricted to global communicators, but there are some use cases (e.g., Sync Batch Norm, #1923), where it can be useful to only perform ops within a node or similar.

We could imagine extending the API with an additional parameter:

```
hvd.allreduce(t, comm=hvd.Local)
```

**Comment by tgaddair** [comment]
cc @romerojosh

**Comment by EiffL** [comment]
I'm very curious about this issue and whether there are plans to implement this features. My use-case is for large-scale model parallelism using the [Mesh TensorFlow framework](https://github.com/tensorflow/mesh) which currently only supports MPI-like collectives on TPU clusters, but not on regular GPU clusters. As far as I can see, TensorFlow and XLA are still missing some of the collective ops required, in particular multi-node alltoall communications, and I'm hoping it might be easier and more flexible to use horovod as the communication backend than adding these ops to the tensorflow code-base.

For an efficient horovod implementation, the only thing that I would need is the ability to perform collective ops selectively over groups/subset of GPUs, so as to be able to implement multi-dimensional computational meshes. Simply exposing local and global communicators would already be fantastic, user defined communicators would be nice but maybe unnecessarily complicated.

Happy to start looking into how to do this if you have a few pointers, I'm still new to the inner workings of the code :-)

**Comment by tgaddair** [comment]
Thanks for suggesting this @EiffL!  I've been really curious about how we might incorporate some ideas from Mesh TensorFlow into Horovod, so glad to see someone else is thinking about this as well.  

Thanks for offering to look into this.  A PR would be much appreciated.

As a starting point, I would suggest taking a look at where we hardcode in the global communicator within the Horovod collective ops.  For example, here's the MPI implementation of `hvd.allreduce` where we hardcode the global communicator: https://github.com/horovod/horovod/blob/master/horovod/common/ops/mpi_operations.cc#L60

For NCCL, we create the communicator ourselves, so we could hypothetically create separate communicators for local and cross groups: https://github.com/horovod/horovod/blob/master/horovod/common/ops/nccl_operations.cc#L61

The main entrypoint for `hvd.allreduce` is going to be here: https://github.com/horovod/horovod/blob/master/horovod/common/ops/nccl_operations.cc#L61

This is where we would want to plumb in a communicator parameter from the user code.

A really good reference implementation that shows all the parts of the code that need to be modified to add a new parameter like this is @romerojosh's recent PR #1949.

Looking forward to chatting more about this.

**Comment by romerojosh** [comment]
I was thinking about this a little bit and wanted to chime in with some thoughts/ideas. In order to support operations across a subset of ranks, we will require handling these communicator subsets in _both_ the control and  data planes. The control logic, at least in the case of the cached responses, expects all workers to have the same set of requests. In order to be able to have only a subset of workers submit a request, we will need to generalize the controller to work on subcommunicators as well. I think this should be fairly straightforward to do if we create multiple controller objects that operate using subcommunicators representing the various "process groups" the user creates.

As a general flow for how I think this could work:
1. User registers a process group via a new API, say something like `proc_group = hvd.create_process_group(ranks=[0,2,4,6])`. This would be called globally by all workers, even those not in the process group so that everyone can keep a consistent ordered list of process groups that exist. This will be important in the next step. Then users can call collectives on the subset like suggested: `hvd.allreduce(tensor, proc_group=proc_group)`. 
2. In `RunLoopOnce`, rather than only processing requests using the single global communicator/Controller from `ComputeReponseList`, the background thread can iterate over a list of `Controller` objects associated with the registered process groups, and generate a full list of responses containing both the normal global collectives, as well as any collective requests for subcommunicators. Since we are careful in step 1 to ensure that these process groups are registered in a globally consistent order, the resulting full response list should be automatically ordered in a way that avoids deadlocking.
3. This full response list is handled as usual by `PerformOperation`. In order to support operations on subcommunicators, we will need to add support to have the data plane backends create and operate on new process group subcommunicators as they are encountered during the run. 

This is just a rough idea but I think in general, an approach like this would be feasible.

**Comment by EiffL** [comment]
Hi there :-) It took a little while... but with a couple of colleagues (@kimchitsigai @mypey) we attempted an implementation of what you described @romerojosh @tgaddair and our prototype appears to be working.

Before opening a PR draft here, we were curious if you might have comments on our approach, and whether this is something that could be of interest and potentially merged at some point into horovod.
We tried to document our approach on this draft PR https://github.com/DifferentiableUniverseInitiative/horovod/pull/2 which you can also use to see what changes we made.

To summarize, this would allow the user to do the following:
```python
from mpi4py import MPI
import horovod.tensorflow as hvd

# This will be our baseline world communicator
comm = MPI.COMM_WORLD
# Split COMM_WORLD into subcommunicators
subcomm = MPI.COMM_WORLD.Split(color=MPI.COMM_WORLD.rank % 2,
                               key=MPI.COMM_WORLD.rank)

# Initialize horovod with an array of communicators
hvd.init(comm=[comm, subcomm])

....

# AlltoAll over the WORLD communicator
hvd.alltoall(a, communicator_id=0)

# AlltoAll over second (splitted) communicator
hvd.alltoall(a, communicator_id=1)
```
Internally this would be handled by having an array of controllers/mpi_contexts/queues, one for each user-specified communicator. This mostly just involves wrapping the existing horovod logic (we inherit from the current horovod support for custom MPI communicators) and as suggested by @romerojosh it's just a matter of looping through the controllers in `RunLoopOnce`. 

This prototype is not extensively tested nor complete, but we did achieve the correct behavior over NCCL. And it's likely that it will suit our needs for a GPU-based Mesh TensorFlow implementation :-)

Could you let us know what you think? We are happy to push this concept to a full PR if you think that's useful.

Many thanks for your help!

**Comment by maxhgerlach** [comment]
@EiffL, it looks like there have been some duplicate effort -- I have been working on a very similar extension in parallel. Unfortunately I didn't learn about your PR before today. Maybe it would be best to join forces?

It looks as if your draft PR is a bit further along in terms of op support, while #2839 (just uploaded) comes with a more dynamic API where communicators don't need to be set at initialization time.

**Comment by EiffL** [comment]
Great to see more people would be interested in having access to this! 100% happy to combine efforts! 
I'll have a careful look at your PR, I'm curious to see what approach you took :-) but at first glance it appears a little bit more involved.

**Comment by romerojosh** [comment]
Excellent work @EiffL and @maxhgerlach! Really glad to see there is so much progress made moving this feature forward. I will try and take a look at these PRs when I get a chance, but I do think merging efforts would be the best way forward. Let us know if there is anything we can do to assist with this.

**Comment by maxhgerlach** [comment]
Hi @EiffL,

it's definitely cool to have so much progress here! We've also missed this feature for quite a while.

I looked over your PR and I really see some advantages of the way you went with split communicator controllers instead of the explicitly enumerated process sets in my PR. As long as every process is part of the same (split) communicator (e.g., either part of the "even" or the "odd" subcommunicator), Horovod does not need to worry about deadlocks or stalls arising from different processes wanting to do different things at the same time and this keeps things a bit simpler. Actually, I am not sure if there are any realistic use cases for not perfectly split communicators. Might there be complications with `hvd.join()` in any way, though?

Another aspect that I like for its simplicity is how you just went with a vector of `OperationManger` objects, each holding its own `MPIContext`. I went about this in a more convoluted way that I don't really enjoy.

Apart from those points, the main thing that's more involved in my attempt is the ability to add and remove process sets dynamically. That required some care to avoid race conditions from local multithreading (not sure if it's 100% clean yet, actually) and to avoid deadlocks from global synchronization. That global synchronization also adds a tiny bit of latency in each Horovod cycle although that's hopefully insignificant. Having this dynamic API may make some applications easier to implement. It's definitely helpful for unit tests where Horovod is initialized just once for dozens of tests. I believe it may not be supported to shutdown an re-initialize MPI in a running process, but I am not sure if there are any gotchas if we don't `MPI_Finalize()` in `hvd.shutdown()`.

Personally, I'd prefer the API to be independent of MPI. That should, for instance, make it easy to port this feature to run with just Gloo in an environment without any MPI. Also it might be less intimidating for users who did not grow up on a Torque or Slurm cluster. ;)

If I may suggest, one way to go about merging these efforts would be to add an entry point `hvd.init(comm=[comm, subcomm])`, which would then set up the process sets corresponding to these externally provided communicators.

At any rate, I'm open to any other ideas how to go forward / suggestions for improvement!

**Comment by maxhgerlach** [comment]
@EiffL, @romerojosh I've added some more commits to #2839. Firstly, I refactored things to hopefully be a bit less convoluted (each process set has its own MPIContext now). Secondly,  Horovod can now be initialized with a static list of process sets, which allows one to get rid of the extra latency to synchronize all processes at the beginning of each Horovod cycle. Additionally, there is a new entry point `hvd.init(comm=[comm, subcomm])` and one can call `hvd.comm_process_set(subcomm)` to get the fitting process set id in each rank. Hopefully, this API should be helpful for users who start out with split MPI communicators.

What do you think?

**Comment by EiffL** [comment]
Very sorry for dropping communications @maxhgerlach but this is really great! Many thanks for your continued efforts on it, I'm going to try out your fork today and tomorrow and do a deep dive into your implementation. I'll let you know if I have any comments or encounter any problems.

**Comment by maxhgerlach** [comment]
@EiffL, that's awesome, let's iron out any remaining quirks! Looking forward to any comment 🙂
